bundles/org.openhab.binding.remehaheating/src/main/java/org/openhab/binding/remehaheating/internal/RemehaHeatingHandler.java

```diff
@@ -0,0 +1,346 @@
+import com.google.gson.JsonObject;
+
+/**
+ * The {@link RemehaHeatingHandler} handles communication with Remeha Home heating systems.
+ * 
+ * This handler manages the connection to the Remeha API, authenticates using OAuth2 PKCE flow,
+ * and provides access to heating system data including temperatures, water pressure, and DHW controls.
+ * 
+ * Supported features:
+ * - Room and outdoor temperature monitoring
+ * - Target temperature control
+ * - Hot water temperature and status monitoring
+ * - DHW mode control (anti-frost, schedule, continuous-comfort)
+ * - Water pressure monitoring
+ * - System status monitoring
+ *
+ * @author Michael Fraedrich - Initial contribution
+ */
+@NonNullByDefault
+public class RemehaHeatingHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(RemehaHeatingHandler.class);
+    private final HttpClient httpClient;
+    private @Nullable RemehaApiClient apiClient;
+    private @Nullable ScheduledFuture<?> refreshJob;
+    private @Nullable RemehaHeatingConfiguration config;
+
+    public RemehaHeatingHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Handles commands sent to the binding channels.
+     * 
+     * Supported commands:
+     * - RefreshType: Updates all channel states from API
+     * - DecimalType on targetTemperature: Sets new target room temperature
+     * - StringType on dhwMode: Changes DHW mode (anti-frost/schedule/continuous-comfort)
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            updateData();
+        } else if (CHANNEL_TARGET_TEMPERATURE.equals(channelUID.getId())) {
+            if (command instanceof QuantityType<?> qt) {
+                QuantityType<?> celsius = qt.toUnit(SIUnits.CELSIUS);
+                if (celsius != null) {
+                    setTargetTemperature(celsius.doubleValue());
+                }
+            } else if (command instanceof DecimalType dt) {
+                setTargetTemperature(dt.doubleValue());
+            }
+        } else if (CHANNEL_DHW_MODE.equals(channelUID.getId()) && command instanceof StringType) {
+            setDhwMode(command.toString());
+        }
+    }
+
+    /**
+     * Initializes the handler by validating configuration and authenticating with Remeha API.
+     * Sets up periodic data refresh job on successful authentication.
+     */
+    @Override
+    public void initialize() {
+        try {
+            config = getConfigAs(RemehaHeatingConfiguration.class);
+            String email = config.email;
+            String password = config.password;
+            int refreshInterval = config.refreshInterval;
+
+            if (email.isEmpty() || password.isEmpty()) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Email and password required");
```

Comment:
Please make use of custom i18n keys when setting the thing state. This makes it possible to add translations.
For reference: https://www.openhab.org/docs/developer/utils/i18n.html#using-custom-keys (last paragraph)